### ISSUE TITLE
Wrong environment detected in unit test under certain conditions

### ISSUE THREAD

**Issue by LMCom** [laravel/framework#25591]
- Laravel Version: 5.4.36
- PHP Version: 7.0.30 (production), 7.1.16 (local)

### Description:
Being very cautious, I implemented a check for the correct environment into the setUp() of my TestCase class.

```php
       if(env('APP_ENV') !== 'testing') {
            $this->fail('Not in testing environment according to APP_ENV. Aborting');
            die(1);
        }

        $status = Artisan::call('config:clear');

        if($status !== 0) {
            $this->fail('Could not clear config cache. Aborting');
            die(1);
        }

        $status = Artisan::call('view:clear');

        if($status !== 0) {
            $this->fail('Could not clear view cache. Aborting');
            die(1);
        }

        $env = app()->environment();

        if($env !== 'testing') {
            $this->fail('App does not detect testing environment via app()->environment(). Detected environment: "' . $env . '". Aborting.');
            die(1);
        }
```

Glad, I did. The first test fails with the following message in my phpunit log:
```xml
<failure type="PHPUnit_Framework_AssertionFailedError">ConfigTest::testConfigCacheClearing
App does not detect testing environment via app()-&gt;environment(). Detected environment: "production". Aborting.

[...]/TestCase.php:41
</failure>
```

This only happens, if I called "php artisan config:clear" immediately prior to the unit test. Subsequent tests do not fail. But this is the lesser important part of the bug. I think, the return value of app()->environment() must never be affected by config:clear. 
As you can see, the value of APP_ENV is checked aswell and has the correct value "testing", otherwise the test would fail.

### Steps To Reproduce:
Use the code above in your TestCase setUp() and try running the test with and without a prior config:clear.

**Comment by laurencei** [comment]
The real problem is you are caching your config on your development box. You should not do this, as it causes problems like what you've described above.

**Comment by LMCom** [comment]
> The real problem is you are caching your config on your development box.

I made the same assumption first. That's why I wrote:

>  This only happens, if I called "php artisan config:clear" immediately prior to the unit test.

So there should be no cached config. Additionally, the unit test itself clears the config cache again, as shown in the above code. Please look more closely at what's happening. How does this fit into your hypothesis, that the config cache was to blame?

**Comment by LMCom** [comment]
It looks, like I need to clarify this further.

> This only happens, if I called "php artisan config:clear" immediately prior to the unit test.

This is not the call of "config:clear", that is done via code (I understand your remark concerning the same cycle). I enter this manually in the command line. The command line prints: "Cache cleared successfully.".
After that I manually start the unit test. How could the config cache still exist at this point?

> You should not cache your config in development. There is simply no point. You should only cache on production.

I do not cache the config in development, btw. But that's not the point of this bug report, anyway.

**Comment by laurencei** [comment]
> I do not cache the config in development, btw. But that's not the point of this bug report, anyway.

So why are you then running `Artisan::call('config:clear');` at all?

**Comment by LMCom** [comment]
I resist the urge to go off-topic. At this point, it seems, you are questioning the legitimacy of this issue.
So let me rephrase this, to make things more clear. This can be answered with yes or no:

_If the config is not cached before the booting of the framework, should `env('APP_ENV')` return the same value as `config('app.env')`?_

- If the answer is **no**, then my issue is obsolete. But I would like an explanation in this case.
- If the answer is **yes**, then there is a bug.

**Comment by laurencei** [comment]
`...should env('APP_ENV') return the same value as config('app.env')?`

It depends on what you have inside config/app.php If you have the default `'env' => env('APP_ENV', 'production'),` - then yes.

But you should **never** call `env()` outside of the config classes. Not even in your tests. It is only designed to be used inside the config system.

Regardless - on a fresh install of Laravel 5.7 I get `testing` as the env using your code above (even if I run `config:clear` before and during the tests).

**Comment by LMCom** [comment]
In the actual system, I only use `env()` in config files.

> Regardless - on a fresh install of Laravel 5.7 I get testing as the env using your code above (even if I run config:clear before and during the tests).

Thanks for this info. In this case, the bug appears less urgent for me. I will upgrade as soon as possible and keep an eye on it. I am all the more curious to what causes it.

As a last note, omitting `view:clear` can lead to exceptions in the worst case. For example, if a custom blade directive was changed or an upgrade changed something. I had two cases, where unit tests with `$this->get('/some/url/')` failed, only because the views needed to be recompiled.

**Comment by laurencei** [comment]
>  For example, if a custom blade directive was changed or an upgrade changed something.

Yeah - fair point 👍

**Comment by LMCom** [comment]
Admittedly, it's a rare case. I told you, I was paranoid. And that got me into that hassle. :)

**Comment by madrzak** [comment]
Accidentally had config cached on my local. Ran tests and it wiped my database. Frustrating as it's not the first time this happened (previously spent a lot of time investigating what could have caused it). So now I added the checks from @LMCom  in my TestCase! 😄 Shouldn't happen again.
